### PR TITLE
fix fade_transition issue

### LINKED ISSUE #157312: Rapidly switching with AnimatedSwitcher causes error:`Failed assertion: line 1191 pos 20: 'node.isRepaintBoundary': is not true.`
### Steps to reproduce

1. Run the sample code.
2. Quickly click the button (log in and log out).

### Expected results

When switching back quickly, there should be no errors, and the animation should continue smoothly

### Actual results

Cause the app to break and freeze.

### Code sample

<details open><summary>Code sample</summary>

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatefulWidget {
  const MyApp({super.key});

  @override
  State<MyApp> createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  bool signedIn = false;

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        colorScheme: ColorScheme.fromSeed(seedColor: Colors.deepPurple),
        useMaterial3: true,
      ),
      home: AnimatedSwitcher(
        duration: const Duration(seconds: 1),
        child: signedIn
            ? HomeScreen(onPressed: () => setState(() => signedIn = false))
            : AuthScreen(onPressed: () => setState(() => signedIn = true)),
      ),
    );
  }
}

class HomeScreen extends StatelessWidget {
  static const routeName = 'homeScreen';

  final VoidCallback onPressed;

  const HomeScreen({super.key, required this.onPressed});

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      key: const ValueKey('homeScreen'),
      appBar: AppBar(
        backgroundColor: Colors.blue,
        title: const Text('Home Screen', style: TextStyle(color: Colors.white)),
      ),
      body: Center(
        child: ElevatedButton(
          onPressed: onPressed,
          style: ButtonStyle(backgroundColor: WidgetStateProperty.all(Colors.blue)),
          child: const Text('log out', style: TextStyle(color: Colors.white)),
        ),
      ),
    );
  }
}

class AuthScreen extends StatelessWidget {
  static const routeName = 'authScreen';

  final VoidCallback onPressed;

  const AuthScreen({super.key, required this.onPressed});

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        backgroundColor: Colors.blue,
        title: const Text('Auth Screen', style: TextStyle(color: Colors.white)),
      ),
      body: Center(
        child: ElevatedButton(
          onPressed: onPressed,
          style: ButtonStyle(backgroundColor: WidgetStateProperty.all(Colors.blue)),
          child: const Text('log in', style: TextStyle(color: Colors.white)),
        ),
      ),
    );
  }
}

```

</details>

### Screenshots or Video

<details open>
<summary>Screenshots / Video demonstration</summary>

https://github.com/user-attachments/assets/76e42e97-7d61-4356-902e-673afd2829c8

</details>

### Logs

_No response_

### Flutter Doctor output

<details open><summary>Doctor output</summary>

```console
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel stable, 3.24.3, on macOS 15.0.1 24A348 darwin-arm64, locale en-NZ)
[✓] Android toolchain - develop for Android devices (Android SDK version 35.0.0)
[✓] Xcode - develop for iOS and macOS (Xcode 16.0)
[✓] Chrome - develop for the web
[✓] Android Studio (version 2024.1)
[✓] IntelliJ IDEA Community Edition (version 2024.1.4)
[✓] VS Code (version 1.77.1)
[✓] Connected device (5 available)
[✓] Network resources

```

</details>

### PR DESCRIPTION
**Pull request by yiiim**
Fixes: #157312

A simpler way to reproduce this issue:
```dart
import 'package:flutter/material.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatefulWidget {
  const MyApp({super.key});

  @override
  State<MyApp> createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> with SingleTickerProviderStateMixin {
  late AnimationController controller = AnimationController(
    duration: const Duration(seconds: 2),
    value: 1,
    vsync: this,
  );
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        body: Center(
          child: FadeTransition(
            opacity: controller,
            child: Builder(
              builder: (context) {
                return GestureDetector(
                  onTap: () {
                    controller.value = 0.5;
                    context.findRenderObject()?.markNeedsPaint();
                    controller.value = 0;
                  },
                  child: Text("Click"),
                );
              },
            ),
          ),
        ),
      ),
    );
  }
}

```
The main reason is that updating the opacity with `markNeedsCompositedLayerUpdate` followed by `markNeedsPaint` causes it to be added to `owner!._nodesNeedingPaint` twice.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

### CODE REVIEW

File: packages/flutter/test/widgets/fade_transition_test.dart [hunk 1]
```diff
@@ -33,4 +34,25 @@ void main() {
     debugPrint = oldPrint;
     debugPrintBuildScope = false;
   });
+
+  // Regression test for https://github.com/flutter/flutter/issues/157312
+  testWidgets('No exception when calling markNeedsPaint after changing the value and then resetting the animated value to 0',
+  (WidgetTester tester) async {
```
Comment (nate-thegrate) [inline packages/flutter/test/widgets/fade_transition_test.dart:40]: nit: maybe the test name could be a bit more concise?

```suggestion
  testWidgets('No exception when calling markNeedsPaint during opacity changes', (WidgetTester tester) async {
```

File: packages/flutter/test/widgets/fade_transition_test.dart [hunk 2]
```diff
@@ -3,14 +3,15 @@
 // found in the LICENSE file.
 
 import 'package:flutter/foundation.dart';
+import 'package:flutter/material.dart';
```
Comment (nate-thegrate) [inline packages/flutter/test/widgets/fade_transition_test.dart:6]: I'm guessing this may have been auto-imported… would it be possible to remove it?

```suggestion
```

**Review by nate-thegrate — APPROVED** [review]
LGTM with a couple of tiny suggestions :)

Thanks for the fix and the well-structured test!

**Comment by yiiim** [comment]
> LGTM with a couple of tiny suggestions :)
> 
> Thanks for the fix and the well-structured test!

Thank you for leaving a comment. I am concerned about the changes to `markNeedsPaint`, and we might need someone more familiar with `RenderObject` to review it.

**Comment by justinmc** [comment]
@dkwingsmt Friendly reminder this needs review now that you're back, if you're up for it.

**Comment by dkwingsmt** [comment]
Thanks for the contribution. The change seems reasonable to me. But since this involves a change to `RenderObject`, I think we should add a test for `RenderObject` so that this fix is verified even if `AnimatedOpacity` is gone.

Also cc @jonahwilliams who worked on related code 3 years ago and should be more knowledgeable than me.


### Side note

It took me a while to understand the relationship between `markNeedsCompositedLayerUpdate` and `markNeedsPaint`:

|  | `markNeedsPaint` | `markNeedsCompositedLayerUpdate` | 
|----|----|----|
| Behavior if it is and was a boundary | Adds itself to `owner!._nodesNeedingPaint` and request visual update | Adds itself to `owner!._nodesNeedingPaint` and request visual update |
| Behavior | (a lot of things) | Same as `markNeedsPaint` |

So basically, other than the first few lines (`if (xxx) {return;} _xxx = true; `), the rest of the two functions are identical. I suggest we refactor it somehow to indicate it more clearly. (It doesn't have to be done in this PR.)

**Review by jonahwilliams — APPROVED** [review]
Change seems reasonable to me, thank you for tracking this one down

LGTM

**Comment by yiiim** [comment]
> But since this involves a change to `RenderObject`, I think we should add a test for `RenderObject` so that this fix is verified even if `AnimatedOpacity` is gone.

@dkwingsmt Tests for `RenderObject` have been added, feel free to review.

**Review by dkwingsmt — APPROVED** [review]
LGTM

File: packages/flutter/lib/src/rendering/object.dart [hunk 3]
```diff
@@ -3041,6 +3041,9 @@ abstract class RenderObject with DiagnosticableTreeMixin implements HitTestTarge
     // If this was not previously a repaint boundary it will not have
     // a layer we can paint from.
     if (isRepaintBoundary && _wasRepaintBoundary) {
+      if (_needsCompositedLayerUpdate) {
```
[the remark this answers is not in the document]
Comment (dkwingsmt) [reply packages/flutter/lib/src/rendering/object.dart:3044]: Well, you caught me forgetting to submit the pending comment. :)

**Comment by nate-thegrate** [comment]
BTW – I've learned that Google testing can have a hard time understanding merge commits, so it might be better to do a rebase here

https://github.com/flutter/flutter/blob/master/docs/contributing/Tree-hygiene.md#using-git

**Comment by yiiim** [comment]
@dkwingsmt Did the Google testing fail because of this assertion?

**Comment by dkwingsmt** [comment]
Yeah it was. But I don't think we can remove this assertion, otherwise it means this change will change the logic of existing apps. Let me see if I can find why it was triggered.

**Comment by dkwingsmt** [comment]
I took a look and propose a different fix:

```
// object.dart, L2964
-       owner?._nodesNeedingPaint.remove(this);
+       owner?._nodesNeedingPaint.removeWhere((RenderObject t) => t == this);
```

The reason why adding the object to `owner!._nodesNeedingPaint` twice causes the problem is because while `flushCompositingBits` is supposed to clean up all objects that have become non-compositing, the job fails because only  the first encounter is removed.

And the reason why I think the current approach might not be the best is because the system is so complicated that we can't assume `_needsCompositedLayerUpdate == true` is equivalent to `owner!._nodesNeedingPaint.contains(this)`. 

And it seems `_nodesNeedingPaint` is designed to allow duplicates, since `flushPaint` checks whether a node still `_needsPaint` on each iteration afresh.

If you use this approach, maybe we should add a comment to `_nodesNeedingPaint` to explain that there might be duplicate nodes.

**Comment by yiiim** [comment]
I have committed the solution you proposed, If `_nodesNeedingPaint` allows duplicates, then we can handle it this way.

File: packages/flutter/lib/src/rendering/object.dart [hunk 4]
```diff
@@ -2958,7 +2958,7 @@ abstract class RenderObject with DiagnosticableTreeMixin implements HitTestTarge
     if (!isRepaintBoundary && _wasRepaintBoundary) {
       _needsPaint = false;
       _needsCompositedLayerUpdate = false;
-      owner?._nodesNeedingPaint.remove(this);
+      owner?._nodesNeedingPaint.removeWhere((RenderObject t) => t == this);
```
Comment (dkwingsmt) [inline packages/flutter/lib/src/rendering/object.dart:2961]: My apology for yet another comment, but I just realized `identical` should be more efficient for not going through `operator ==`.
```suggestion
      owner?._nodesNeedingPaint.removeWhere((RenderObject t) => identical(t, this));
```

**Comment by dkwingsmt** [comment]
@jonahwilliams Would you like to take another look, since we've changed the approach?

**Review by dkwingsmt — APPROVED** [review]
LGTM
